src/packbuilder.rs

```diff
@@ -0,0 +1,263 @@
+
+use {raw, panic, Repository, Error, Oid, Revwalk, Buf};
+use util::{Binding};
+
+pub enum Stage {
+    AddingObjects,
+    Deltafication,
+}
+
+type ProgressCb<'a> = FnMut(Stage, u32, u32) -> bool + 'a;
+type ForEachCb<'a> = FnMut(&[u8]) -> bool + 'a;
+
+/// TODO(Nemo157): Documentation
+pub struct PackBuilder<'repo> {
+    raw: *mut raw::git_packbuilder,
+    _marker: marker::PhantomData<&'repo Repository>,
+}
+
+impl<'repo> PackBuilder<'repo> {
+    /// TODO(Nemo157): Documentation
+    pub fn insert_object(&mut self, id: Oid, name: Option<&str>) -> Result<(), Error> {
+        let name = try!(::opt_cstr(name));
+        unsafe {
+            try_call!(raw::git_packbuilder_insert(self.raw, &*id.raw(), name));
+        }
+        Ok(())
+    }
+
+    /// TODO(Nemo157): Documentation
+    pub fn insert_tree(&mut self, id: Oid) -> Result<(), Error> {
+        unsafe {
+            try_call!(raw::git_packbuilder_insert_tree(self.raw, &*id.raw()));
+        }
+        Ok(())
+    }
+
+    /// TODO(Nemo157): Documentation
+    pub fn insert_commit(&mut self, id: Oid) -> Result<(), Error> {
+        unsafe {
+            try_call!(raw::git_packbuilder_insert_commit(self.raw, &*id.raw()));
+        }
+        Ok(())
+    }
+
+    /// TODO(Nemo157): Documentation
+    pub fn insert_walk(&mut self, walk: &mut Revwalk) -> Result<(), Error> {
+        unsafe {
+            try_call!(raw::git_packbuilder_insert_walk(self.raw, walk.raw()));
+        }
+        Ok(())
+    }
+
+    /// TODO(Nemo157): Documentation
+    pub fn insert_recursive(&mut self, id: Oid, name: Option<&str>) -> Result<(), Error> {
+        let name = try!(::opt_cstr(name));
+        unsafe {
+            try_call!(raw::git_packbuilder_insert_recur(self.raw, &*id.raw(), name));
+        }
+        Ok(())
+    }
+
+    /// TODO(Nemo157): Documentation
+    pub fn write_buf(&mut self, buf: &mut Buf) -> Result<(), Error> {
+        unsafe {
+            try_call!(raw::git_packbuilder_write_buf(buf.raw(), self.raw));
+        }
+        Ok(())
+    }
+
+    /// TODO(Nemo157): Documentation
+    pub fn object_count(&self) -> usize {
+        unsafe { raw::git_packbuilder_object_count(self.raw) as usize }
+    }
+
+    /// TODO(Nemo157): Documentation
+    pub fn written(&self) -> usize {
+        unsafe { raw::git_packbuilder_written(self.raw) as usize }
+    }
+
+    /// TODO(Nemo157): Documentation
+    pub fn foreach<F>(&mut self, mut cb: F) -> Result<(), Error>
+                  where F: FnMut(&[u8]) -> bool {
+        let mut cb: &mut ForEachCb = &mut cb;
+        let ptr = &mut cb as *mut _;
+        unsafe {
+            try_call!(raw::git_packbuilder_foreach(self.raw, foreach_c, ptr as *mut _));
+        }
+        Ok(())
+    }
+
+    /// TODO(Nemo157): Documentation
+    pub fn set_progress_callback<'a, F>(&'a mut self, mut cb: &'a mut F) -> Result<(), Error>
+                         where F: FnMut(Stage, u32, u32) -> bool {
+        let ptr = &mut cb as *mut _;
```

Comment:
Hm I think this may not be sound? This passes the data payload as a stack pointer to libgit2, but that data is invalid after this function returns?

This may want to store `Box<FnMut(..)>` internally and then pass that boxed pointer down to libgit2